superchain/superchain.go

```diff
@@ -351,30 +388,66 @@ func (c ContractImplementations) Resolve(versions ContractVersions) (Implementat
 	if implementations.SystemConfig, err = resolve(c.SystemConfig, versions.SystemConfig); err != nil {
 		return implementations, fmt.Errorf("SystemConfig: %w", err)
 	}
+	// If the L2OutputOracle is not specified (versions.L2OutputOracle is empty), we can assume that the L2OutputOracle
+	// is not used, and fault proofs are activated.
+	if implementations.L2OutputOracle, err = resolve(c.L2OutputOracle, versions.L2OutputOracle); errors.Is(err, ErrEmptyVersion) {
+		if implementations.SystemConfig, err = resolve(c.SystemConfig, versions.SystemConfig); err != nil {
+			return implementations, fmt.Errorf("SystemConfig: %w", err)
+		}
+		if implementations.AnchorStateRegistry, err = resolve(c.AnchorStateRegistry, versions.AnchorStateRegistry); err != nil {
+			return implementations, fmt.Errorf("AnchorStateRegistry: %w", err)
+		}
+		if implementations.DelayedWETH, err = resolve(c.DelayedWETH, versions.DelayedWETH); err != nil {
+			return implementations, fmt.Errorf("DelayedWETH: %w", err)
+		}
+		if implementations.DisputeGameFactory, err = resolve(c.DisputeGameFactory, versions.DisputeGameFactory); err != nil {
+			return implementations, fmt.Errorf("DisputeGameFactory: %w", err)
+		}
+		if implementations.FaultDisputeGame, err = resolve(c.FaultDisputeGame, versions.FaultDisputeGame); err != nil {
+			return implementations, fmt.Errorf("FaultDisputeGame: %w", err)
+		}
+		if implementations.MIPS, err = resolve(c.MIPS, versions.MIPS); err != nil {
+			return implementations, fmt.Errorf("MIPS: %w", err)
+		}
+		if implementations.PermissionedDisputeGame, err = resolve(c.PermissionedDisputeGame, versions.PermissionedDisputeGame); err != nil {
+			return implementations, fmt.Errorf("PermissionedDisputeGame: %w", err)
+		}
+		if implementations.PreimageOracle, err = resolve(c.PreimageOracle, versions.PreimageOracle); err != nil {
+			return implementations, fmt.Errorf("PreimageOracle: %w", err)
+		}
+	} else if err != nil {
+		return implementations, fmt.Errorf("L2OutputOracle: %w", err)
+	}
 	return implementations, nil
 }
 
 // resolve returns a VersionedContract that matches the passed in semver version
 // given a set of addresses.
 func resolve(set AddressSet, version string) (VersionedContract, error) {
```

Comment:
Updated the way `resolve` functions here.  We now just perform exact matching to resolve the `version` against the list of implementations (`set`).

Previously, an empty `version` would be compared to each version in `set`, and the underlying `semver` package would [silently sort this invalid version as "less than"](https://cs.opensource.google/go/x/mod/+/master:semver/semver.go;l=126) the valid versions.  This caused an empty version to be fulfilled with the latest implementation.